### ISSUE TITLE
Extra secure paths that Nix cannot modify

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#7471]
**Is your feature request related to a problem? Please describe.**

We are getting closer to #5208, a nix daemon without root, but an issue that is brought up is that this currently doesn't provide much value on NixOS: Since the system itself is running via symlinks to the store, the nix daemon could in principle modify those paths and take over the system giving itself root's powers.

**Describe the solution you'd like**

A simple and ham-fisted solution is that store paths in the NixOS closure should continue to be owned by root. The daemon can in fact not manage them and will fail if it tries to do so.

Nix itself can't and shouldn't know about NixOS, so from it's perspective it would look like this:
 - all paths in the Nix Store must be *at least* as secure as the nix daemon user/group/whatever owning them.
 - some paths, for reasons unknown to the Nix daemon, may be *more secure*. So it would need to understand that having the owner be some random unknown user is *bad* and a security hole, vs having it root (I guess view some UID/GID or whitelist? Or just hard-code UID 0 as a special case?) is *not bad* because even though the practical effect that the daemon cannot manage those files is the same, it trusts those other more powerful users to take care of those files.

Then we just need Nix to fail gracefully when trying to do things with root-owned paths.

The `nixos-rebuild` script or similar will have to manage the `chown`ing of paths to and from root (from root when the nixos generation is removed and thus those paths become free to be garbage collected).

**Describe alternatives you've considered**

Not sure, I welcome other ideas.

**Additional context**

See #5208 for more details of what the rootless nix daemon is planned to look like.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by thufschmitt** [comment]
I'm not sure that this really makes sense. If nothing else, it's the daemon that creates this “extra-secure” objects, so it can already do whatever. So it seems that this wouldn't add anything security-wise because there's no workaround to trusting the daemon if you run things from the Nix store.

**Comment by Ericson2314** [comment]
I was thinking daemon wouldn't have perms to chown to root, so something else would do it.

**Comment by thufschmitt** [comment]
> I was thinking daemon wouldn't have perms to chown to root, so something else would do it.

Yes, obviously. But that something else would `chown` a store path that the daemon created, right?

**Comment by Ericson2314** [comment]
Sure, but I think that is OK. It will not guard against past malicious stuff, but it *will* guard against future shenanigans.

FWIW I am less worried about an evil Nix daemon than a build that manages to escape its sandbox; against that threat makes sense to secure these paths if we think a machine hasn't been compromised so far, one one is going to build arbitrary stuff in the future and who knows what will happen then.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
